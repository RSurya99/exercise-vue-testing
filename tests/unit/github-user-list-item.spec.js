import GithuUserListItem from '@/components/exercise/github-user-list-item.vue'
import { mount } from '@vue/test-utils'

let wrapper
const userData = {
  login: 'RSurya99',
  id: 73375663,
  node_id: 'MDQ6VXNlcjczMzc1NjYz',
  avatar_url: 'https://avatars.githubusercontent.com/u/73375663?v=4',
  gravatar_id: '',
  url: 'https://api.github.com/users/RSurya99',
  html_url: 'https://github.com/RSurya99',
  followers_url: 'https://api.github.com/users/RSurya99/followers',
  following_url: 'https://api.github.com/users/RSurya99/following{/other_user}',
  gists_url: 'https://api.github.com/users/RSurya99/gists{/gist_id}',
  starred_url: 'https://api.github.com/users/RSurya99/starred{/owner}{/repo}',
  subscriptions_url: 'https://api.github.com/users/RSurya99/subscriptions',
  organizations_url: 'https://api.github.com/users/RSurya99/orgs',
  repos_url: 'https://api.github.com/users/RSurya99/repos',
  events_url: 'https://api.github.com/users/RSurya99/events{/privacy}',
  received_events_url: 'https://api.github.com/users/RSurya99/received_events',
  type: 'User',
  site_admin: false,
  name: 'Rafli Surya P',
  company: null,
  blog: '',
  location: 'Bandung, Indonesia',
  email: null,
  hireable: null,
  bio: null,
  twitter_username: 'RSurya99',
  public_repos: 29,
  public_gists: 0,
  followers: 38,
  following: 68,
  created_at: '2020-10-24T04:06:24Z',
  updated_at: '2022-01-07T01:23:43Z',
}

describe('Github User List Item', () => {
  beforeEach(() => {
    wrapper = mount(GithuUserListItem, {
      propsData: {
        user: userData,
      },
    })
  })
  it('duplicate the card of specific user when the Duplicate button is clicked', () => {
    const button = wrapper.find('[data-testid=duplicateButton]')

    button.trigger('click')
    const duplicateCalls = wrapper.emitted('duplicate')
    expect(duplicateCalls).toHaveLength(1)
    const expectedPayload = { user: userData }
    expect(wrapper.emitted('duplicate')[0][0]).toMatchObject(expectedPayload)
  })
  it('remove the card of specific user when the Remove button is clicked', () => {
    const button = wrapper.find('[data-testid=removeButton]')

    button.trigger('click')
    // 4. Assert event has been emitted
    const removeCalls = wrapper.emitted('remove')
    expect(removeCalls).toHaveLength(1)
    const expectedPayload = { user_id: 73375663 }
    expect(wrapper.emitted('remove')[0][0]).toMatchObject(expectedPayload)
  })
})
