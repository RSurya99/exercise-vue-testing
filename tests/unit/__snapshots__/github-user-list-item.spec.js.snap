// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Github User List Item renders correctly 1`] = `
<li class="gh-list-item">
  <div class="gh-list-item__actions"><button data-testid="duplicateButton">Duplicate</button> <button data-testid="removeButton">Remove</button></div>
  <figure class="gh-list-item__profpic">
    <div class="gh-list-item__profpic__inner"><img src="https://avatars.githubusercontent.com/u/73375663?v=4" alt="profile picture"></div>
  </figure>
  <div>
    <header>
      <h3 style="margin: 0px;">RSurya99</h3>
      <p style="margin: 0px; font-size: 0.8rem;">Bandung, Indonesia</p>
    </header>
    <section class="gh-list-item__section-info">
      <p class="gh-list-item__section-info__item"><span>Public Repos</span> <span>29</span></p>
      <p class="gh-list-item__section-info__item"><span>Followers</span> <span>38</span></p>
      <p class="gh-list-item__section-info__item"><span>Following</span> <span>68</span></p>
    </section>
    <section class="gh-list-item__section-info">
      <p class="gh-list-item__section-info__item"><span>Registered at</span> <span>10-24-2020</span></p>
    </section>
  </div>
</li>
`;
